frontend/src/app/messages/messages.service.ts

```diff
@@ -0,0 +1,91 @@
+import { Injectable } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
+import { Observable } from 'rxjs/Observable';
+import { Subject } from 'rxjs/Subject';
+import { environment } from '../../environments/environment';
+
+
+export interface ServerMessage {
+  message: string;
+  type: string;
+}
+
+export class Message {
+  message: string;
+  type: string;
+  dismissible: boolean;
+  /** Timeout is set in seconds, not ms. Can be floating-point. */
+  timeout?: number;
+}
+
+export enum MessageType {
+  Success = 'success',
+  Info = 'info',
+  Warning = 'warning',
+  Danger = 'danger',
+  Primary = 'primary',
+  Secondary = 'secondary',
+  Light = 'light',
+  Dark = 'dark'
+}
+
+export const DEFAULT_TIMEOUT_TIME = 7.5;
+
+@Injectable()
+export class MessagesService {
+  private messageSubject = new Subject<Message>();
+  public message$ = this.messageSubject.asObservable();
+
+  constructor(private http: HttpClient) { }
+
+  private nextMessage(message?: Message) {
+    this.messageSubject.next(message);
+  }
+
+  showMessage(message: string, type = MessageType.Info, dismissible = true,
+    timeout = DEFAULT_TIMEOUT_TIME) {
+    this.nextMessage({ message, type, dismissible, timeout });
+  }
+
+  success(message: string, dismissible = true, timeout = DEFAULT_TIMEOUT_TIME) {
+    this.showMessage(message, MessageType.Success, dismissible, timeout);
+  }
+
+  info(message: string, dismissible = true, timeout = DEFAULT_TIMEOUT_TIME) {
+    this.showMessage(message, MessageType.Info, dismissible, timeout);
+  }
+
+  warning(message: string, dismissible = true, timeout = DEFAULT_TIMEOUT_TIME) {
+    this.showMessage(message, MessageType.Warning, dismissible, timeout);
+  }
+
+  danger(message: string, dismissible = true, timeout = DEFAULT_TIMEOUT_TIME) {
+    this.showMessage(message, MessageType.Danger, dismissible, timeout);
+  }
+
+  fetchServerMessages() {
+    this.http.get<ServerMessage[]>(`${environment.apiRoot}/messages/`)
+      .subscribe(serverMessages => {
+        serverMessages
+          .map(this.mapServerMessage)
+          .forEach(this.nextMessage.bind(this));
+      });
+  }
+
+  private mapServerMessage(serverMessage: ServerMessage): Message {
+    const { message, type } = serverMessage;
+    return {
+      message,
+      // Change message type when there is no match with Django message type.
+      // Object.values() logic is a workaround around the 'in' operator working
+      // improperly on string Enums in TypeScript. Maybe they will fix it someday...
+      type: Object.values(MessageType).includes(type) ? type : MessageType.Warning,
```

Comment:
There is a mismatch between here and showMessage. showMessage function falls back to Info type and here you fall back to Warning type. Is it intended?

Reply:
It was intended, as it was a generalization of logic invented by magul when he created a component for the first time. It mapped sever 'error' type to boostrap 'warning' type. However, there was also a case with 'debug' messages from backend that was not covered, hence I generalised it. However, with new changes a made the fallback to INFO type.